cpp/src/arrow/device.cc

```diff
@@ -193,6 +193,9 @@ Result<std::shared_ptr<Buffer>> CPUMemoryManager::ViewBufferFrom(
   if (!from->is_cpu()) {
     return nullptr;
   }
+  if (buf->device_type() != DeviceAllocationType::kCPU) {
```

Comment:
Perhaps add a comment motivating this piece of code?

Reply:
added